### PR TITLE
statement level highlighting

### LINKED ISSUE #44: Statement level highlighting
## Description

Each method's coverage entry should expose expose executed vs non-executed statements via plain accessors.

## Acceptance Criteria
- [ ] Test of expected statement list if the accessors.

## Extras
- Minimal inspector extension

## Estimate

1 day

---

### LINKED ISSUE #45: Docs
## Description

Update the README with this Epic example usage and test the profiler fully

## Acceptance Criteria
- [ ] Write some E2E test for the profiler

## Testing Notes

Repeated execution should be idempotent.

## Estimate

0.5 day

---

### PR DESCRIPTION
**Pull request by renman-ymd**
## Summary
This pull request introduces several improvements and corrections to the code coverage instrumentation framework, focusing on more accurate coverage reporting, improved test coverage for edge cases, and consistent naming.
**Ordered coverage result:** The internal counter dictionary in coverage results is now an `OrderedIdentityDictionary`, preserving statement order for more predictable highlighting and reporting.

### Documentation Updates

- **Readme update:** The `README.md` now includes a section on code coverage, with usage examples, coverage queries, and statement highlighting.
- **Benchmark report:** A new benchmarking report (`docs/REPORT.md`) details the methodology, workloads, scenarios, and findings for instrumentation performance and coverage profiling.

Closes #44
Closes #45

### CODE REVIEW

File: docs/REPORT.md [hunk 1]
```diff
@@ -0,0 +1,257 @@
...
+
+2. **H1 ≫ A1/R1 by ~16×** (0.082 vs 0.005 µs/firing). The cost is in
+   the `IGCounter` action body itself (Dictionary work) more than in the
+   instrumentation send. A collector that just appends a value to an
+   `OrderedCollection` is 16× cheaper per firing than a dictionary
+   counter.
+
+3. **The `messageSendHeavy:` C1 outlier (0.59 µs/firing)** suggests
+   per-firing cost depends on AST-node shape. Coarse expressions =
+   bigger reified literals = slower.
```
Comment (guillep) [inline docs/REPORT.md:155]: Is this due to the hashing of AST nodes? otherwise I don't see how the "size" of the node affects speed.

Unless we are talking about instrumentation granularity.
So, two methods written in different ways will have different per-statement instrumentations.

File: docs/REPORT.md [hunk 2]
```diff
@@ -0,0 +1,257 @@
...
+| `tinyMethod` | `^ 42` | 200 000 calls |
+| `lightMethod` | 5 assignments + return | 100 000 calls |
+| `mediumMethod` | 20 assignments + return | 30 000 calls |
+| `heavyMethod` | `x:=0` + 50 increments + return | 10 000 calls |
+| **`straightLine200`** | 200 OrderedCollection sends | 1 000 calls |
+| `assignmentChain` | `a := b := c := d := 42` | 100 000 calls |
+| `tightLoop:` | `whileTrue:` with 1-stmt body | 500 000 iters |
+| `tightLoopHeavy:` | `whileTrue:` with 16-stmt body | 30 000 iters |
+| `messageSendHeavy:` | self-send loop | 200 000 sends |
+| `recursiveFactorial:` | recursive | depth 500 |
```
Comment (guillep) [inline docs/REPORT.md:34]: Would be nice to have the code.
Assignments do not work the same if they are temp vars, inst vars or global vars.
Specially regarding dead code elimination.
For example, instance variable assignments cannot be simply be (re)moved.

File: docs/REPORT.md [hunk 3]
```diff
@@ -0,0 +1,257 @@
...
+
+Several B0 measurements come in absurdly low because Cog's JIT eliminates
+their bodies (return value unused, no observable side effect). For
+example, `mediumMethod` baseline reads as 0.0074 µs/call, which is below
+the cost of a single object allocation. That's the JIT erasing the body,
+not real performance.
+
+When a baseline is DCE-affected, use the **absolute per-firing cost**, not
+the slowdown ratio. The slowdown ratio reflects the JIT compiler more
+than Insight.
```
Comment (guillep) [inline docs/REPORT.md:69]: Well, in some way we can say that Insight prevents the JIT compiler optimization! So the slowdown is real.

The question is, how many methods do actually have a shape like that in real life...

File: docs/REPORT.md [hunk 4]
```diff
@@ -0,0 +1,257 @@
...
+
+### 2.2 Hook reach: most hooks reach inside blocks; R1 does not
+
+A probe revealed that **R1** (Collector capturing `statementResult`)
+only fires on top-level statements.
+
+R1's restriction is a property of `IGStatementResultReification>>#rewriteOperationsForNode:`,
+which only wraps statements at the top level
```
Comment (guillep) [inline docs/REPORT.md:77]: So this is a bug, would you mind open an issue with an example?

File: docs/REPORT.md [hunk 5]
```diff
@@ -0,0 +1,257 @@
...
+20 measured iterations, on `tightLoopHeavy:` the collector
+grows to 10M+ entries:
+
+| Stat | A1 | A2 | R1 |
+|---|---|---|---|
+| min | 856 825 µs | 2 008 371 µs | 3 689 470 µs |
+| median | 2 050 836 µs | 3 354 145 µs | 3 810 986 µs |
+| max | 3 686 179 µs | 5 645 076 µs | 3 850 126 µs |
+
+The 4× span between A1 min and max is GC pressure.
```
Comment (guillep) [inline docs/REPORT.md:91]: Nice, so accummulating data in an ordered collection may not be the most efficient.

My only issue is that this are very artificial micro benchmarks.
Can we have an estimate of memory consumption on some realisting workload like test case coverage?

File: docs/REPORT.md [hunk 6]
```diff
@@ -0,0 +1,257 @@
...
+
+† JIT-DCE on baseline (§2.1): per-firing computed against an unrealistically
+fast B0, so the absolute number is somewhat distorted but proportionally
+small (the `tightLoop:` 0.165 underestimates by a few percent).
+
+**Headline: a C1 hook firing costs approximately 0.21–0.27 µs across
+clean workloads.** The `messageSendHeavy:` outlier at 0.59 µs/firing is
+because each "firing" there happens on a more complex AST node (a self-send
+expression with arguments rather than an assignment), which produces a
+larger reified `astNode` literal.
```
Comment (guillep) [inline docs/REPORT.md:124]: See my comment below

File: docs/REPORT.md [hunk 7]
```diff
@@ -0,0 +1,257 @@
...
+   bigger reified literals = slower.
+
+#### Per-call overhead
+
+For a method with N statements, the per-call overhead is
+roughly `N × 0.215 µs` for C1.
+
+If a method uninstrumented body takes ~10 µs, C1 instrumentation costs **~40% slowdown**.
+If the body takes 100 µs, **~4% slowdown**. If the body takes 1 µs
+(e.g. a accessor), **~430% slowdown**.
```
Comment (guillep) [inline docs/REPORT.md:164]: I don't understand these :)

per-call overhead is per-call to an instrumented method, or per-call to a hook?
Of course the overhead is relative to the number of hooks installed and what those hooks do.
But these estimates are assuming something that is not explicit and I don't get

File: docs/REPORT.md [hunk 8]
```diff
@@ -0,0 +1,257 @@
...
+
+---
+
+## 4. Hot-spot analysis
+
+### 4.1 Per-firing cost decomposes into three layers
+
+```
+Bare hook firing (literal load + 1 send):       ~0.005 µs
++ A1 capture work (OrderedCollection add):      negligible
```
Comment (guillep) [inline docs/REPORT.md:213]: So we could have a collector that just introduces an add to a collection, right? removing the intermediate calls

File: docs/REPORT.md [hunk 9]
```diff
@@ -0,0 +1,257 @@
...
++ Counter dictionary work:                       ~0.077 µs
++ Counter[astNode] hash + bigger keyset:         ~0.13 µs
```
Comment (guillep) [inline docs/REPORT.md:215]: cost of computing the hash? Very probably hash is implemented recursively.
Otherwise I don't see why such a difference.

File: docs/REPORT.md [hunk 10]
```diff
@@ -0,0 +1,257 @@
...
+```
+
+### 4.2 `Dictionary at:update:initial:` is the biggest win available
+
+The 0.13 µs/firing is the largest single-line item. Replacing
+the per-node Dictionary in `IGCounter` with something like a per-method indexed array
+(populated at instrument time, indexed by treePath) would reduce the cost by an estimate of **60%**.
```
Comment (guillep) [inline docs/REPORT.md:222]: This would be a nice alternative counter implementation, yes!

File: docs/REPORT.md [hunk 11]
```diff
@@ -0,0 +1,257 @@
...
+
+### 4.3 Collector unbounded growth is a usability issue
+
+`tightLoopHeavy:` A1 went 1.24 s (min) → 6.55 s (max) across 20 iterations.
+The collector's `OrderedCollection` grew to ~10M entries. **Per-firing
+CPU cost is fine**; the wall-clock blow-up is GC.
+
```
Comment (guillep) [inline docs/REPORT.md:229]: We should keep this in the back of our heads, but I'd say this is minor until we find a real case that showcases this issue

**Review by guillep — APPROVED** [review]
Thanks @renman-ymd ! The analysis was fun to read!

I think this is ok to go, we can iterate in my comments later

File: docs/REPORT.md (hunk 2 above)
Comment (renman-ymd) [reply docs/REPORT.md:34]: I can push a branch with the generated code if you want
